### ISSUE TITLE
syncthing -generate=blah doesn't generate a device ID with existing certs

### ISSUE THREAD

**Issue by ColinChartier** [syncthing/syncthing#4217]
To test:
```
syncthing
mkdir /tmp/derp
cp ~/.config/syncthing/*.pem /tmp/derp/
syncthing -generate=/tmp/derp
cat /tmp//derp/config.xml | grep "device id"
        <device id="" introducedBy=""></device>
    <device id="" name="5d9a0362b945" compression="metadata" introducer="false" skipIntroductionRemovals="false" introducedBy="">

```

**Comment by AudriusButkevicius** [comment]
I don't think it's supposed to. Why do you think it's supposed to?

**Comment by calmh** [comment]
You skipped the output from Syncthing, which I think shows quite clearly that this is intentional:

```
jb@unu:~ $ syncthing -generate=foo
07:43:10 INFO: Device ID: FZ23CS4-PDV743V-32LS44M-TCIUBAQ-RQ3OC4X-XN6EG66-BGUCZH4-OISTMAE
07:43:10 INFO: Default folder created and/or linked to new config
jb@unu:~ $ syncthing -generate=foo
07:43:12 WARNING: Key exists; will not overwrite.
07:43:12 INFO: Device ID: FZ23CS4-PDV743V-32LS44M-TCIUBAQ-RQ3OC4X-XN6EG66-BGUCZH4-OISTMAE
07:43:12 WARNING: Config exists; will not overwrite.
jb@unu:~ $ 
```

**Comment by ColinChartier** [comment]
@calmh Sorry, I'm still confused-- why is Device ID generated and printed from the first call, but never put into a config and/or retrievable?

I'm trying to get the device ID from a specific public key (that I have generated on my own) to template it into configurations myself.

How can I get the device ID besides parsing the stdout of syncthing -generate=(the folder)?

**Comment by AudriusButkevicius** [comment]
The code that generates it is here:

https://github.com/syncthing/syncthing/blob/master/lib/protocol/deviceid.go

Essentially it's sha256 hash of the certificate, base32 encoded, equals trailers stripped, and luhn check digits added.

**Comment by calmh** [comment]
The device ID is saved into the config if it's generated. Apparently that doesn't happen if the key already exists and the config doesn't - that's a corner case we've never tested/considered, I think. Regardless, it'll be fixed on first startup.

Parsing -generate output is a perfectly fine way of detecting the device ID of a given key, I think.

**Comment by ColinChartier** [comment]
I'm just worried that "log statements aren't an API" will burn me at some later point if you decide to change the log format or something, and I can't duplicate your algorithm to generate device ID as that has already changed in the past (to add the check digits) -- is there a possibility of getting a `syncthing deviceid cert.pem` command or equivalent to freeze the API?

**Comment by AudriusButkevicius** [comment]
You can run syncthing with those keys, query the rest api and kill it.
You are spawning a subprocess anyway.

**Comment by calmh** [comment]
Here's a one liner for you:

```
openssl x509 -in cert.pem -outform der \
  | openssl sha -sha256 -binary \
  | python -c 'import base64; import sys; print(base64.b32encode(sys.stdin.read()))' \
  | tr -d =
```

That gets you the device ID, without dashes and check digits. Nonetheless this format is accepted by Syncthing, in the config and in the GUI.

**Comment by rmoriz** [comment]
@calmh would you be so kind adding that openssl example to the documentation? (https://docs.syncthing.net/dev/device-ids.html)  near 

> An advanced user could replace the key.pem and cert.pem files with a keypair generated directly by the openssl utility or other mechanism.

Thank you!

**Comment by calmh** [comment]
No. See the OpenSSL documentation for how to use OpenSSL, or other tool documentation for how to use other tools. We will not usefully duplicate it.

On 1 Sep 2017, at 22:52, rmoriz <notifications@github.com<mailto:notifications@github.com>> wrote:


@calmh<https://github.com/calmh> would you be so kind adding that openssl example to the documentation? (https://docs.syncthing.net/dev/device-ids.html) near

An advanced user could replace the key.pem and cert.pem files with a keypair generated directly by the openssl utility or other mechanism.

Thank you!

**Comment by rmoriz** [comment]
So why did you even add `openssl x509` example to the documentation? That's just inconsistent. If you want to make it hard, just remove all hints.

**Comment by calmh** [comment]
That's an example of how to look at the keys. It's necessary to illustrate the article. 

If you want to generate keys using OpenSSL instead of letting Syncthing manage it it's presumably because you have specific requirements. Only you can know what they are and how to fulfill them and any example we might provide would be wrong. The times I've needed to generate keys with OpenSSL I've googled a tutorial like everyone else, it's not something I know by heart. 

In short, go do your own research.

**Comment by rmoriz** [comment]
I need to verify if a given key pair matches a device ID. IMHO syncthing should provide a cli option that just returns the device id based on the parsed key pair and does not start the main application. (think of `syncthing -identity`).

As you pointed out in this issue above, it's not possible to retrieve the device id just with common X.509 tooling because (by design) syncthing does not use X.509 subject information and additionally calculates/adds "check digits" outside of the traditional X.509 workflow.

With an "-identity" option and the already implemented "-generate" functionality, syncthing can manage itself and also be deployed using typical DevOps-Tooling.
